src/hotspot/share/logging/logAsyncWriter.hpp

```diff
@@ -140,7 +55,79 @@ typedef ResourceHashtable<LogFileStreamOutput*,
 // change the logging configuration via jcmd, LogConfiguration::configure_output() calls flush() under the protection of the
 // ConfigurationLock. In addition flush() is called during JVM termination, via LogConfiguration::finalize.
 class AsyncLogWriter : public NonJavaThread {
+  friend class AsyncLogTest;
   class AsyncLogLocker;
+  using AsyncLogMap = ResourceHashtable<LogFileStreamOutput*,
+                          uint32_t,
+                          17, /*table_size*/
+                          ResourceObj::C_HEAP,
+                          mtLogging>;
+  class Message {
+    LogFileStreamOutput* const _output;
+    const LogDecorations _decorations;
+
+   public:
+    // only-valid for in-place new!
+    Message(LogFileStreamOutput* output, const LogDecorations& decorations, const char* msg)
+      : _output(output), _decorations(decorations) {
+      assert(msg != nullptr, "c-str message can not be null!");
+      PRAGMA_STRINGOP_OVERFLOW_IGNORED
+      strcpy(reinterpret_cast<char* >(this+1), msg);
+    }
+
+    size_t size() const {
+      size_t len = strlen(message()) + 1;
+      return align_up(sizeof(*this) + len, sizeof(void*));
+    }
+
+    bool is_token() const { return _output == nullptr; }
+    LogFileStreamOutput* output() const { return _output; }
+    const LogDecorations& decorations() const { return _decorations; }
+    const char* message() const { return reinterpret_cast<const char *>(this+1); }
+  };
+
+  // A flush TOKEN is Message(nullptr, None, "");
+  static const size_t TOKEN_SIZE = {align_up(sizeof(Message) + 1, sizeof(void*))};
+
+  class Buffer : public CHeapObj<mtLogging> {
+    size_t _pos;
+    char* _buf;
+    size_t _capacity;
+
+   public:
+    Buffer(size_t capacity);
+    ~Buffer();
+
+    void push_flush_token();
+    bool push_back(LogFileStreamOutput* output, const LogDecorations& decorations, const char* msg);
+
+    // for testing-only!
+    size_t set_capacity(size_t value) {
+      size_t old = _capacity;
+      _capacity = value;
+      return old;
+    }
```

Comment:
Actually, I would prefer it if you did not shrink the buffer dynamically, since this is not a supported scenario and may cause unforeseen problems (besides making the code harder to argue about). 

You could just leave buffer size immutable after creation and instead add another test variation to your `AsyncLogGtest.java` where you start with a tiny tiny buffer size, and then run all your gtests with it. More test coverage, and we don't need the test coding here.

(Side note, I see that in AsyncLogGtest.java you run all tests inside a single `@test` scope. If you split them up and give them speaking id's, these tests can run in parallel. For examples, see e.g. the NMTGtests. The duplication of the test section is annoying, but the parallelization of tests you get for free is cool.)

Reply:
if we want to test 'tiny-buffer' elegantly, we have to lift the constraint of `AsyncLogBufferSize`. I just feel that we shouldn't leave a door for testing so end-user can mess with it. 

Code like this is a compromise. I think it's fine because Buffer isn't designed as a reusable component. We might be over-engineering if we try to tide up everything. What do you think?